issue 332 - rocksdb skip options Destroy()

steveyen opened the pull request:
Proposing that leaking the writeoptions right now is a better tradeoff / workaround than having Close() crash for folks who are using rocksdb.

mschoch commented:
Steve, I didn't test any code, but this does not line up with my understanding.  The merge operator destroy which is causing the crash is set in the DB-level options object.  So, if we're going to leak anything to avoid a crash, it would need to be that.  Though, I would propose we get the problem fixed upstream and not leak.

steveyen commented:
Hi Marty,
I _think_ this PR is doing what you're thinking of.  Here's my understanding and summary...

If a bleve user who has a blevex/rocksdb store invokes index.Close(), there's a crash, as index.Close() calls store.Close(), which calls writeoptions.Destroy(), which crashes.  That's seen from the stack traces reported on https://github.com/blevesearch/bleve/issues/332.

To help debug, I added a PR to bleve-blast, where I added an index.Close() at the end of bleve-blast, and can reproduce: https://github.com/blevesearch/bleve-bench/pull/11 ...so, I'm able to get the same crash stack traces.

In this PR to blevex/rocksdb, it's a leaky workaround... just no longer call writeoptions.Destroy() during rocksdb store.Close().  And, after that, bleve-blast no longer crashes.

> Though, I would propose we get the problem fixed upstream and not leak.

Also agree to also doing that, in where I don't mind a leaky interim workaround until upstream mystery is solved..
